compiler/grainlsp/hover.re

```diff
@@ -0,0 +1,247 @@
+    switch (value_bindings) {
+    | [] => ("Internal error: empty let", None)
+    | _ =>
+      let matches =
+        List.map(
+          (vb: Typedtree.value_binding) =>
+            Utils.get_node_from_expression(~line, ~char, vb.vb_expr),
+          value_bindings,
+        );
+      let filtered =
+        List.filter(
+          m =>
+            switch ((m: Utils.node_t)) {
+            | Error(_) => false
+            | NotInRange => false
+            | _ => true
+            },
+          matches,
+        );
+      switch (filtered) {
+      | [] =>
+        let vb = List.hd(value_bindings);
+        let expr = vb.vb_expr;
+        let pat = vb.vb_pat;
+        (
+          Utils.expression_lens(~line, ~char, ~compiled_code, expr),
+          Some(pat.pat_loc),
+        );
+
+      | [node] =>
+        switch (node) {
+        | Error(err) => (err, None)
+        | NotInRange => ("Not in range", None)
+        | Expression(e) => (
+            Utils.expression_lens(~line, ~char, ~compiled_code, e),
+            if (e.exp_loc == Grain_parsing.Location.dummy_loc) {
+              Some(stmt.ttop_loc);
+            } else {
+              Some(e.exp_loc);
+            },
+          )
+        | Pattern(p) => (
+            Utils.markdown_grain(
+              Utils.lens_sig(p.pat_type, ~env=compiled_code.env),
+            ),
+            if (p.pat_loc == Grain_parsing.Location.dummy_loc) {
+              Some(stmt.ttop_loc);
+            } else {
+              Some(p.pat_loc);
+            },
+          )
+        }
+      | _ => ("Internal error", None)
+      };
+    }
+
+  | TTopExpr(expression) =>
+    // Because of the List syntax sugar, we don't get a location
+    // But as a cover all, if we have a dummy location here,
+    // use the enclosing expression location
+
+    let (loc, node) =
+      if (expression.exp_loc == Grain_parsing.Location.dummy_loc) {
+        let nd: Utils.node_t = Expression(expression);
+        (stmt.ttop_loc, nd);
+      } else {
+        (
+          expression.exp_loc,
+          Utils.get_node_from_expression(~line, ~char, expression),
+        );
+      };
+
+    switch (node) {
+    | Error(err) => ("Error: " ++ err, None)
+    | NotInRange => (
+        Utils.markdown_grain(
+          Utils.lens_sig(expression.exp_type, ~env=compiled_code.env),
+        ),
+        Some(loc),
+      )
+    | Expression(e) => (
+        Utils.expression_lens(~line, ~char, ~compiled_code, e),
+        if (e.exp_loc == Grain_parsing.Location.dummy_loc) {
+          Some(loc);
+        } else {
+          Some(e.exp_loc);
+        },
+      )
+
+    | Pattern(p) => (
+        Utils.markdown_grain(
+          Utils.lens_sig(p.pat_type, ~env=compiled_code.env),
+        ),
+        Some(p.pat_loc),
+      )
+    };
+
+  | TTopException(export_flag, type_exception) =>
+    let exception_type = type_exception.ext_type; // Not sure how to get a good value here
```

Comment:
The type of all exceptions is just `Exception`, but you could show the type of this specific constructor.